From bc7bc9c951be8f8d84421b0d91cf45d820b5d63e Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sun, 16 Aug 2026 16:27:33 -0400
Subject: [PATCH 273/285] winewayland: Do not lock wayland pointer when
 accessing frame.

frame state is only touched in the dispatcher thread, so we don't need to lock.
---
 dlls/winewayland.drv/wayland_pointer.c | 18 ++----------------
 1 file changed, 2 insertions(+), 16 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_pointer.c b/dlls/winewayland.drv/wayland_pointer.c
index 7145e78ee04..5badfb61ae8 100644
--- a/dlls/winewayland.drv/wayland_pointer.c
+++ b/dlls/winewayland.drv/wayland_pointer.c
@@ -210,9 +210,10 @@ static void pointer_handle_enter(void *data, struct wl_pointer *wl_pointer,
     pthread_mutex_lock(&pointer->mutex);
     pointer->focused_hwnd = hwnd;
     pointer->enter_serial = serial;
-    wayland_pointer_reset_frame();
     pthread_mutex_unlock(&pointer->mutex);
 
+    wayland_pointer_reset_frame();
+
     /* The cursor is undefined at every enter, so we set it again with
      * the latest information we have. */
     wayland_set_cursor(hwnd, NULL, FALSE);
@@ -296,7 +297,6 @@ static void pointer_handle_axis(void *user_data, struct wl_pointer *wl_pointer,
 
     if (!(hwnd = wayland_pointer_get_focused_hwnd())) return;
 
-    pthread_mutex_lock(&pointer->mutex);
     switch (axis)
     {
         case WL_POINTER_AXIS_VERTICAL_SCROLL:
@@ -309,7 +309,6 @@ static void pointer_handle_axis(void *user_data, struct wl_pointer *wl_pointer,
             break;
         default: break;
     }
-    pthread_mutex_unlock(&pointer->mutex);
 
     TRACE("hwnd %p scroll %.3f\n", hwnd, scroll);
 }
@@ -326,9 +325,6 @@ static void pointer_handle_frame(void *data, struct wl_pointer *wl_pointer)
     TRACE("hwnd %p\n", hwnd);
 
     input.type = INPUT_MOUSE;
-
-    pthread_mutex_lock(&pointer->mutex);
-
     input.mi.dwFlags = MOUSEEVENTF_ABSOLUTE | MOUSEEVENTF_MOVE;
 
     if (frame->flags & WAYLAND_POINTER_FRAME_ABSOLUTE)
@@ -400,7 +396,6 @@ static void pointer_handle_frame(void *data, struct wl_pointer *wl_pointer)
 
     frame->flags = 0;
     frame->scroll = frame->horz_scroll = 0;
-    pthread_mutex_unlock(&pointer->mutex);
 }
 
 static void pointer_handle_axis_source(void *data, struct wl_pointer *wl_pointer,
@@ -417,7 +412,6 @@ static void pointer_handle_axis_stop(void *data, struct wl_pointer *wl_pointer,
 
     if (!(hwnd = wayland_pointer_get_focused_hwnd())) return;
 
-    pthread_mutex_lock(&pointer->mutex);
     switch (axis)
     {
         case WL_POINTER_AXIS_HORIZONTAL_SCROLL:
@@ -428,7 +422,6 @@ static void pointer_handle_axis_stop(void *data, struct wl_pointer *wl_pointer,
             break;
         default: break;
     }
-    pthread_mutex_unlock(&pointer->mutex);
 
     TRACE("hwnd %p axis %u\n", hwnd, axis);
 }
@@ -443,7 +436,6 @@ static void pointer_handle_axis_value120(void *data, struct wl_pointer *wl_point
     if (!value120) return;
     if (!(hwnd = wayland_pointer_get_focused_hwnd())) return;
 
-    pthread_mutex_lock(&pointer->mutex);
     switch (axis)
     {
         case WL_POINTER_AXIS_VERTICAL_SCROLL:
@@ -456,7 +448,6 @@ static void pointer_handle_axis_value120(void *data, struct wl_pointer *wl_point
             break;
         default: break;
     }
-    pthread_mutex_unlock(&pointer->mutex);
 
     TRACE("hwnd=%p axis=%u value120=%d\n", hwnd, axis, value120);
 }
@@ -549,17 +540,12 @@ static void relative_pointer_v1_relative_motion(void *private,
         raw_y = screen_y;
     }
 
-    pthread_mutex_lock(&pointer->mutex);
-
     frame->dx_raw += raw_x;
     frame->dy_raw += raw_y;
     frame->dx += screen_x;
     frame->dy += screen_y;
-
     frame->flags |= WAYLAND_POINTER_FRAME_RELATIVE;
 
-    pthread_mutex_unlock(&pointer->mutex);
-
     TRACE("hwnd=%p screen=%.2f,%.2f raw=%.2f,%.2f accum=%.2f,%.2f\n",
           hwnd, screen_x, screen_y, raw_x, raw_y, frame->dx_raw, frame->dy_raw);
 }
-- 
2.55.0

